File: push/src/main/java/io/appmetrica/analytics/push/impl/notification/NotificationStatus.java
package io.appmetrica.analytics.push.impl.notification;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.util.Collections;
import java.util.Set;

public class NotificationStatus {

    public static class Group {

        @NonNull
        public final String id;
        @NonNull
        public final Set<Channel> channels;
        public final boolean enabled;
        public final boolean changed;

        public Group(@NonNull String id, @Nullable Set<Channel> channels, boolean enabled, boolean changed) {
            this.id = id;
            this.enabled = enabled;
            this.channels = channels == null ? Collections.<Channel>emptySet() : Collections.unmodifiableSet(channels);
            this.changed = changed;
        }
    }

    public static class Channel {

        @NonNull
        public final String id;
        public final boolean enabled;
        public final boolean changed;

        public Channel(@NonNull String id, boolean enabled, boolean changed) {
            this.id = id;
            this.enabled = enabled;
            this.changed = changed;
        }
    }

    @NonNull
    public final Set<Group> groups;
    @NonNull
    public final Set<Channel> channelsWithoutGroup;
    public final boolean enabled;
    public final boolean changed;
    @Nullable
    private Long changedTime;

    public NotificationStatus(final boolean enabled, final boolean changed) {
        this(Collections.<Group>emptySet(), Collections.<Channel>emptySet(), enabled, changed);
    }

    public NotificationStatus(@NonNull final Set<Group> groups,
                              @NonNull final Set<Channel> channelsWithoutGroup,
                              final boolean enabled,
                              final boolean changed) {
        this.groups = Collections.unmodifiableSet(groups);
        this.channelsWithoutGroup = Collections.unmodifiableSet(channelsWithoutGroup);
        this.enabled = enabled;
        this.changed = changed;
    }

    @Nullable
    public Long getChangedTime() {
        return changedTime;
    }

    public void setChangedTime(@Nullable Long changedTime) {
        this.changedTime = changedTime;
    }
}

File: push/src/main/java/io/appmetrica/analytics/push/impl/notification/NotificationStatus.kt
package io.appmetrica.analytics.push.impl.notification

import org.json.JSONObject

internal class NotificationStatus @JvmOverloads constructor(
    val enabled: Boolean,
    val changed: Boolean,
    val groups: Set<NotificationStatusGroup> = setOf(),
    val channelsWithoutGroup: Set<NotificationStatusChannel> = setOf()
) {

    private val enabledKey = "enabled"
    private val changedKey = "changed"
    private val groupsKey = "groups"
    private val channelsWithoutGroupKey = "channels"
    private val systemNotifyTimeKey = "system_notify_time"

    private var changedTime: Long? = null

    fun getChangedTime() = changedTime

    fun setChangedTime(changedTime: Long?) {
        this.changedTime = changedTime
    }

    fun toJson() =
        JSONObject().apply {
            put(enabledKey, enabled)
            put(systemNotifyTimeKey, changedTime)
            put(changedKey, if (changed) true else null)

            if (groups.isNotEmpty()) {
                put(
                    groupsKey,
                    JSONObject().apply {
                        for (group in groups) {
                            put(group.id, group.toJson())
                        }
                    }
                )
            }

            if (channelsWithoutGroup.isNotEmpty()) {
                put(
                    channelsWithoutGroupKey,
                    JSONObject().apply {
                        for (channel in channelsWithoutGroup) {
                            put(channel.id, channel.toJson())
                        }
                    }
                )
            }
        }
}
